A bunch of unit tests and a Dummy implementation

stain opened the pull request:
This mainly is intended to test out if the API makes sense, and to give examples of what it can be used for on its own. I found several ambiguities that I will raise individually.

The dummy implementation is basically just some beans and a java.util.List. This uses the Java 8 Streaming APIs, so I did not attempt to back-patch everything to java6 - the `java6` profile simply ignores the whole of `src/test/java` for now.

The tests are written through two abstract classes, so that they can be reused as-is in a different implementation. The test-jar would probably be needed for that (See #40)

The only thing needed to provide is an instance of the [RDFTermFactory](https://github.com/stain/commons-rdf/blob/tests/src/test/java/com/github/commonsrdf/api/RDFTermFactory.java) - I put this in `src/test/java` for now - although it might be genuinely useful as part of the APIs. (I have not tested how easy it is to plug into this from other implementations).

There's a jewel test in there that you would enjoy if you like SPARQL...

stain commented on src/test/java/com/github/commonsrdf/dummyimpl/BlankNodeImpl.java:
```diff
@@ -0,0 +1,53 @@
+		if (id == null || id.isEmpty() || id.contains(":")) {
+			// TODO: Check against
+			// http://www.w3.org/TR/n-triples/#n-triples-grammar
+			throw new IllegalArgumentException("Invalid blank node id: " + id);
```
Well, it was meant as a 'reference implementation' that can potentially move out of src/test (which is kind of "private until you explicitly depend on tests"). But I agree on your point and therefore rather throw a IllegalStateException from `ntriplesString` if there is a `:` in the id.

stain commented on src/test/java/com/github/commonsrdf/api/AbstractRDFTermFactoryTest.java:
```diff
@@ -0,0 +1,237 @@
+		assertEquals("http://example.com/vocab#term", term.getIRIString());
+		assertEquals("<http://example.com/vocab#term>", term.ntriplesString());
+
+		IRI relative = factory.createIRI("../relative");
```
Relative IRIs are defined in http://www.w3.org/TR/rdf11-concepts/#section-IRIs - but I agree that implementations might not support this, thus it should not be a REQUIREMENT to pass the tests. Given that we don't have a "capabilities" profile or a possibility for a junit test to give a warning, I now conditionally skip these tests if `factory.createIRI("../relative")` throws an exception (and hinted at using IllegalArgumentException in the factory).

ansell replied:
It refers to relative IRIs in the context of concrete syntaxes only so it wouldn't be a requirement. Could you change the factory.createIRI method to require a base IRI to be passed into the method along with the relative IRI?

stain commented on src/test/java/com/github/commonsrdf/api/AbstractRDFTermFactoryTest.java:
```diff
@@ -0,0 +1,237 @@
+				ntriplesString.startsWith("_:"));
+		assertTrue("Internal identifier can't be empty", bnode
+				.internalIdentifier().length() > 0);
+		assertEquals("ntriples does not correspond with internal identifier",
```
I'm not sure I'm with you. The result of `ntriplesString()` should surely be a valid ntriples string for a blank node? If it can't create one I have thrown an `IllegalStateException` - perhaps this could be hinted at in the interface of `RDFTerm.ntriplesString()`?

Does this mean that you can also have an `IRI` which can't give a valid `ntriplesString`?

Are you saying that if an implementation decides to support non-compatible bnode identifiers (say one containing `:`), then it is also free to use such in the auto-generated identifier of `factory.createBlankNode()` - and hence have default blank nodes which can't respond to `ntriplesString()` method? This sounds like bad practice and exactly something this unit test should detect.

If you really want to support that (which I would not like..) - then we should make the whole ntriplesString method optional with `throws UnsupportedOperationException` hint.

ansell replied:
The internal identifiers that may be used are not limited to the N-Triples subset, but that may not be clear in the current contract. Hence, the point I was trying to make was that this test, which is being designed as a compliance test for all commons-rdf implementations, should not assume that the internal identifiers are either preserved or limited to the ntriples subset.

As you say, it would probably be better to either remove ntriplesString or make it an optional method via the unsupported operation route for this reason.

stain replied:
Could you raise the ntriplesString-is-optional as a separate issue and then we can adjust the test accordingly? (e.g. to also catch `UnsupportedOperationException` for `ntriplesString()`)

I found it one of the most useful methods..  :-)

afs replied:
The internal identifier is not the syntax for bNodes in a file.  It maybe used to generate one and an impementation may even internal use a form that is usable directly but it is not required.  The internal identifier is unique (it is in correspondence to the set of blank nodes). It may be mapped for output uses:(e.g. choose short-file-scoped "_:b0", "_:b1", ...

To quote RDF 1.1 : "IRIs, literals and **blank nodes** are distinct and distinguishable."  My emphasis.

That means it could, in the API contract, it usefully aligns to the java equality contract for RDF terms.

stain replied:
Agree - this was fixed in #55.

We still assume that you can always call `ntriplesString` - so the only conclusion now is that the ntriplesString string should be the same if creating with the same internal identifier twice (even if "invalid").

afs replied:
See #56. I think `ntripleString` as a concept in the general common API has problems.

ansell commented on src/test/java/com/github/commonsrdf/api/AbstractGraphTest.java:
```diff
@@ -0,0 +1,192 @@
+				.startsWith(
+						"<http://example.com/alice> <http://xmlns.com/foaf/0.1/name> \"Alice\" ."));
+		assertTrue(graph.toString().endsWith(
+				"_:org1 <http://xmlns.com/foaf/0.1/name> \"A company\" ."));
```
This needs to be changed to avoid referencing the blank node by identifier here, as serialisers and third-party implementations are free to remap them as necessary. (See other comment about checking/relying on n-triples blank node identifiers.

You could move this test to DummyGraphTest to test its behaviour is consistent with its goals and implementation, but not include it in this abstract test that others will run.

ansell commented on src/test/java/com/github/commonsrdf/api/AbstractRDFTermFactoryTest.java:
```diff
@@ -0,0 +1,349 @@
+		// http://www.w3.org/TR/rdf11-concepts/#section-IRIs
+		// it is not a requirement for an implementation to support
+		// it (all instances of an relative IRI should eventually
+		// be possible to resolve to an absolute IRI)
```
They are only defined in the context of concrete syntaxes, so we do not need to support them here.

ansell commented on src/test/java/com/github/commonsrdf/api/AbstractRDFTermFactoryTest.java:
```diff
@@ -0,0 +1,349 @@
+		assertEquals("../relative#term", relativeTerm.getIRIString());
+		assertEquals("<../relative#term>", relativeTerm.ntriplesString());
+
+		IRI emptyRelative = factory.createIRI(""); // <> equals the base URI
```
We don't currently have a base IRI that this could be resolved against when it is being processed as an IRI in memory, similar to other comments about relative IRIs.

stain replied:
There *could* be a base URI depending on how the factory was created - which is out of scope for us. But this test bails out early if it is not supported, without raising it as an error. If it is supported, though, then the rest should work.

Jena's `IRIFactory` for instance can be created with a base URI (and in fact, there even an IRIImpl is a IRIFactory) :)

afs replied:
Yes - a factory implementation choice could be to "auto-resolve" any relative IRI strings it is given.

Relative IRIs are going to be a source of confusion anyway (almost a perma-issue!) - for example, in LDP, it is desirable to create RDF graphs having relative IRIs so that the syntax generated has relative IRIs against an unknonw base.  The base is applied when received at an LDP server.

Such graphs are technically illegal RDF - RDF is only defined for absolute IRIs. 

One approach is to be strict - abolute IRIs only. 

Another is to leave relative IRIs alone because this is a common interface and different code on top may wish to do different things.

stain replied:
Agree - but it's out of scope for us to define exactly where relative IRIs become absolute, and so can't REQUIRE relative IRIs to be unsupported. If they are supported (as tested on the top), then the following tests should be true.

It is certainly useful for an API to be able to generate triple statements which are relative (and don't express a `@base`), say because you don't know the `@base` in advance or are going to store the RDF file on a USB stick. Exactly how to do this varies with implementations - but I must admit that it can still be a bit of a hack for both Jena and Sesame. But all praise the `app://192db400-dc97-4dbf-b613-3741196b4a67/` URI scheme! :=)

afs replied:
Agreed - it can't be required. Can't test the string back from getIRIString to the same as the factory input on relatve URIs.

Actually, not on abolute ones either - it miught have fixed up "/../" or other canonicalizations.

ansell commented on src/test/java/com/github/commonsrdf/api/AbstractRDFTermFactoryTest.java:
```diff
@@ -0,0 +1,349 @@
+		}
+
+		// NOTE: We do not require object equivalence after insertion,
+		// but the ntriples should match
```
The contract will need to be clarified to avoid this implication. As with the other cases, you can put this test in DummyRDFTermFactoryTest to clarify that implementation, but it isn't a general test.

stain replied:
Agree that relying on ntriplesString being preserved is fragile here. It should probably not test any equivalence on the blank node. Fixed in #55.

ansell commented on src/test/java/com/github/commonsrdf/api/AbstractRDFTermFactoryTest.java:
```diff
@@ -0,0 +1,349 @@
+	}
+
+	@Test(expected = Exception.class)
+	public void invalidBlankNode() throws Exception {
```
As above, move to DummyRDFTermFactoryTest.

ansell commented on src/test/java/com/github/commonsrdf/api/AbstractRDFTermFactoryTest.java:
```diff
@@ -0,0 +1,349 @@
+	@Test(expected = Exception.class)
+	public void invalidBlankNode() throws Exception {
+		try {
+			factory.createBlankNode("with:colon").ntriplesString();
```
Just a general note on the style of this line, .ntriplesString is never expected to be called, so easier not to call it.

ansell replied:
Hmm, seems I may have misunderstood this line. ntriplesString is intended to be called and throw an exception, but the test will fail if createBlankNode throws an exception before .ntriplesString().

In that case, it is best to call ntriplesString outside of this try block to avoid confusion in a similar fashion to me.

stain replied:
I updated `invalidBlankNode` in #55 to `possiblyInvalidBlankNode` to do as you say. It now only asserts that if you get that far, then `with:colon` should not be in the ntriplesString.

ansell commented on src/test/java/com/github/commonsrdf/api/AbstractRDFTermFactoryTest.java:
```diff
@@ -0,0 +1,349 @@
+		} catch (UnsupportedOperationException ex) {
+			Assume.assumeNoException("createBlankNode(String) not supported",
+					ex);
+			return;
```
I haven't seen this style before, when used with Test(expected = Exception.class).

Lately I have been using ExpectedException instead of the annotations, which may be a little clearer in this case.

stain replied:
Updated in #55 `possiblyInvalidBlankNode` to avoid confusion (now not expecting any error).
It also tests that if it's OK to get an ntriplesString from the 'invalid' ID - then it's equals if you do it twice.

ansell commented on src/test/java/com/github/commonsrdf/api/AbstractRDFTermFactoryTest.java:
```diff
@@ -0,0 +1,349 @@
+			factory.createLiteral("Example", "");
+		} catch (UnsupportedOperationException ex) {
+			Assume.assumeNoException(
+					"createLiteral(String,String) not supported", ex);
```
This shouldn't be an optional operation. It is required for all implementations for basic functionality.

ansell commented on src/test/java/com/github/commonsrdf/api/AbstractRDFTermFactoryTest.java:
```diff
@@ -0,0 +1,349 @@
+	}
+
+	@Test(expected = Exception.class)
+	public void invalidTriplePredicate() {
```
We still have an open (kind of) discussion about supporting Generalized RDF, where this would be valid, but we would probably use another interface or method to support Generalized RDF so this test is useful as is.

ansell commented on src/test/java/com/github/commonsrdf/api/RDFTermFactory.java:
```diff
@@ -0,0 +1,78 @@
+ * @see Graph
+ * 
+ */
+public interface RDFTermFactory {
```
The conventions around default methods are still not clear, but I don't think we should be implementing them all with UnsupportedOperationException if they are not optional methods that would be required for minimum functionality.

stain replied:
I don't know yet which of these should be "truly" optional - perhaps only `createGraph()`?  Could you raise a new issue about which should and shouldn't be UnsupportedOperationException? Reducing this set would simplify tests as well - currently you can get away with a factory that can't make anything, which is silly.

ansell commented on src/test/java/com/github/commonsrdf/dummyimpl/LiteralImpl.java:
```diff
@@ -0,0 +1,112 @@
+			throw new NullPointerException("literal can't be null");
+		}
+		this.lexicalForm = literal;
+		this.languageTag = Optional.of(languageTag);
```
Could you check the syntax of languageTag against BCP47 using Locale.forLanguageTag?

http://docs.oracle.com/javase/tutorial/i18n/locale/create.html

ansell commented on src/test/java/com/github/commonsrdf/dummyimpl/BlankNodeImpl.java:
```diff
@@ -0,0 +1,58 @@
+
+import com.github.commonsrdf.api.BlankNode;
+
+public class BlankNodeImpl implements BlankNode {
```
Could you change all of the Impl classes that won't be called by external users to be package private to avoid exposing them in the public API? We are not relying on OSGi or Java 9 Modules/etc. where this wouldn't be an issue, but we can avoid exposing them where possible by not making them public.

ansell commented on src/test/java/com/github/commonsrdf/dummyimpl/GraphImpl.java:
```diff
@@ -0,0 +1,133 @@
+public class GraphImpl implements Graph {
+
+	private static final int TO_STRING_MAX = 10;
+	protected List<Triple> triples = new ArrayList<Triple>();
```
Should this be a LinkedHashSet to avoid duplicates?

stain replied:
Yes - it was - but for this to make sense then `Triple` and `IRI` and so on should implement `hashCode()` and `equals()` - that's a separate issue.   

As the javadoc for `Triple` does not require that, I do not implement it here - similarly `Graph.add()` is not required to do anything special on duplicates (and I test covering for both cases, e.g. `afterCount > beforeCount` after multiple add of same triple.

ansell replied:
Although there may be various implementation differences related to SPARQL-1.1 and duplicates from queries, at the abstract RDF-1.1 level Graph is defined to be a set so we shouldn't need to keep duplicates in a reference implementation.

http://www.w3.org/TR/rdf11-concepts/#data-model

I don't think there is any reason for the abstract test to be testing afterCount > beforeCount for duplicates. I missed that bit when I went through. If you know you are working with a List backed Graph then you could add that test in for that implementation to verify its specific implementation.

ansell commented on src/test/java/com/github/commonsrdf/dummyimpl/TestWritingGraph.java:
```diff
@@ -0,0 +1,76 @@
+public class TestWritingGraph {
+
+	/** Run tests with -Dkeepfiles=true to inspect /tmp files **/
+	private static boolean KEEP_FILES = Boolean.getBoolean("keepfiles");
```
This is a useful pattern that I hadn't seen before!

stain replied:
Yes, I like looking at the files.. but don't want to fill up with 100x8 MB on every build :) (just 2x8MB here)

If I do a lot of them I usually make a createTempFile() method in the base test class that does this kind of logic. (Is `keepfiles` good enough name?)

ansell replied:
keepfiles works, it is only exposed at build time so don't need to be too concerned about it.

stain commented:
I added defaults in `RDFTermFactory` that throw `UnsupportedOperationExceptions` and modified `AbstractRDFTermFactoryTest` to `Assume`-bail out early if a method is not supported, e.g. `createGraph()`. 

This means that the test ironically all pass if you don't implement anything.. (see `DefaultRDFTermFactoryTest`)

Similarly `AbstractGraphTest` is OK as long as you implement at least `createGraph()` and `createIRI(String)`) in the factory.

ansell commented on src/test/java/com/github/commonsrdf/dummyimpl/TripleImpl.java:
```diff
@@ -0,0 +1,54 @@
+import com.github.commonsrdf.api.RDFTerm;
+import com.github.commonsrdf.api.Triple;
+
+public class TripleImpl implements Triple {
```
Could you add hashCode and equals implementations here and in the RDFTerm implementations (Eclipse autogenerated versions would be okay)? (See above comment about LinkedHashSet for why this would be useful even in a test set)

stain commented:
Added documentation to https://github.com/stain/commons-rdf/tree/tests#testing

stain commented:
as discussed in [JENA-827](https://issues.apache.org/jira/browse/JENA-827) this can be dangerous as the ISO 639-3 - but testing for that particular `vls` seems to work, so I've added that in 789fc11 and c000685.

The test is changed to test for `"with spaces"` which causes the exception - are we OK to require this to fail for all implementations? (considering you want to allow arbitrary bnode identifiers :) )

Note that you have to `new Locale.Builder().setLanguageTag()` as otherwise Locale will silently swallow the exception.

ansell commented:
Was the content of this pull request merged in #52?

stain commented:
Right, it was merged in #52. I'll close this after going through your comments a second time :)

stain commented:
I've gone through the comments in here, most are addressed by #55. I'll let @ansell suggest which of `RDFTermFactory` methods should be required as a separate issue. So finally closing this!
